src/core/lib/iomgr/work_serializer.cc

```diff
@@ -37,46 +37,35 @@ class WorkSerializer::WorkSerializerImpl : public Orphanable {
  public:
   void Run(std::function<void()> callback,
            const grpc_core::DebugLocation& location);
-
+  void Schedule(std::function<void()> callback,
+                const grpc_core::DebugLocation& location);
+  void DrainQueue();
   void Orphan() override;
 
  private:
-  void DrainQueue();
-
   // An initial size of 1 keeps track of whether the work serializer has been
   // orphaned.
   std::atomic<size_t> size_{1};
+  std::atomic<bool> draining_{false};
   MultiProducerSingleConsumerQueue queue_;
 };
 
 void WorkSerializer::WorkSerializerImpl::Run(
```

Comment:
I think we want to keep the original implementation of `Run()` the way it was instead of implementing it in terms of `Schedule()` and `DrainQueue()`, because it has a performance optimization that is probably important: for the case where no other thread is currently holding the `WorkSerializer`, it will immediately invoke the callback instead of pushing it onto the queue, which avoids a memory allocation.

Reply:
Done. Reworked to avoid the memory allocation for the case when the callback can be invoked immediately.